Gu.Wpf.DataGrid2D/Source2D.Sources.cs

```diff
@@ -76,6 +76,35 @@ private static void OnColumnHeadersChanged(DependencyObject d, DependencyPropert
             {
                 return;
             }
+            
+            //Subscribe to Colection changed events
+              var action = new NotifyCollectionChangedEventHandler(
+                    (o, args) =>
+                    {
+                      var dg = (DataGrid)d;
+                      var rows = (IEnumerable)o;
+
+                      if (dg != null)
+                      {
+                        UpdateColumns(rows, dataGrid);
+                      }
+                    });
+
+            if (e.OldValue != null)
+            {
+              var coll =  e.OldValue as INotifyCollectionChanged ;
```

Comment:
Right! I was puzzled for a minutes there because I thought THAT was what I did. I need to sleep.